viamonvif/service.go

```diff
@@ -131,14 +154,123 @@ func (dis *rtspDiscovery) DiscoverResources(ctx context.Context, extra map[strin
 	return cams, nil
 }
 
+func (dis *rtspDiscovery) DoCommand(ctx context.Context, command map[string]interface{}) (map[string]interface{}, error) {
+	cmd, ok := command["command"].(string)
+	if !ok {
+		return nil, errors.New("invalid command type")
+	}
+
+	switch cmd {
+	case "preview":
+		dis.logger.Debugf("snapshot command received")
+		previewReq, err := toPreviewCommand(command)
+		if err != nil {
+			return nil, err
+		}
+		dis.mu.RLock()
+		snapshotURI, found := dis.RTSPToSnapshotURIs[previewReq.rtspURL]
+		dis.mu.RUnlock()
+		if !found {
+			return nil, fmt.Errorf("snapshot URI not found for %s", previewReq.rtspURL)
+		}
+		dis.logger.Infof("snapshot URI: %s", snapshotURI)
+
+		dataURL, err := downloadPreviewImage(ctx, dis.logger, snapshotURI)
+		if err != nil {
+			return nil, fmt.Errorf("failed to download preview image: %w", err)
+		}
+		return map[string]interface{}{
+			"preview": dataURL,
+		}, nil
+
+	default:
+		return nil, fmt.Errorf("unknown command: %s", cmd)
+	}
+}
+
 func (dis *rtspDiscovery) Close(_ context.Context) error {
 	dis.mdnsServer.Shutdown()
 	return nil
 }
 
+func toPreviewCommand(command map[string]interface{}) (*previewRequest, error) {
+	attributes, ok := command["attributes"].(map[string]interface{})
+	if !ok {
+		return nil, errors.New("attributes is missing or not a map")
+	}
+	rtspURL, ok := attributes["rtsp_address"].(string)
+	if !ok {
+		return nil, errors.New("rtsp_address cannot be empty")
+	}
+	return &previewRequest{rtspURL: rtspURL}, nil
+}
+
+// formatDataURL formats the image data and content type into a data URL.
+func formatDataURL(contentType string, imageBytes []byte) string {
+	base64Image := base64.StdEncoding.EncodeToString(imageBytes)
+	return fmt.Sprintf("data:%s;base64,%s", contentType, base64Image)
+}
+
+// downloadPreviewImage downloads the preview image from the snapshot uri and returns it as a data URL.
+func downloadPreviewImage(ctx context.Context, logger logging.Logger, snapshotURI string) (string, error) {
+	parsedURL, err := url.Parse(snapshotURI)
```

Comment:
we had a URI struct to start with... why not store the (already validated) uri so you don't need to parse it?

Reply:
Which URI struct are you referring to?

MediaInfo holds URIs as strings so that we can easily write as json in the discovery cmd executable.

Reply:
the one returned from GetSnapshotURI